mitiq/zne/scaling/layer_scaling.py

```diff
@@ -0,0 +1,115 @@
+# Copyright (C) 2023 Unitary Fund
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+"""Functions for layer-wise unitary folding on supported circuits."""
+from copy import deepcopy
+import numpy as np
+from typing import Callable, List
+import cirq
+from cirq import Moment, inverse
+
+from mitiq import QPROGRAM
+from mitiq.zne.scaling.folding import _check_foldable
+from mitiq.interface import noise_scaling_converter
+from mitiq.utils import (
+    _append_measurements,
+    _pop_measurements,
+)
+
+
+@noise_scaling_converter
+def layer_folding(
+    circuit: cirq.Circuit, layers_to_fold: List[int]
+) -> cirq.Circuit:
+    """Applies a variable amount of folding to select layers of a circuit.
+
+    Args:
+        circuit: The input cirq circuit.
+        layers_to_fold: A list with the index referring to the layer number,
+                        and the element filled by an integer represents the
+                        number of times the layer is folded.
+
+    Returns:
+        A cirq ``Circuit`` with layers and number of times to fold specified
+        by ``layers_to_invert``.
+    """
+    circuit = cirq.synchronize_terminal_measurements(circuit)
```

Comment:
```suggestion
    circuit = cirq.synchronize_terminal_measurements(circuit)
```
This seems dangerous since it could mutate the input variable `circuit`. But, if it doesn't, ignore this comment.

Reply:
Right. I think this did make sense before but perhaps now can be taken out. Removed!